### ISSUE TITLE
Error response from daemon: Ports are not available: exposing port TCP 0.0.0.0:4505 -> 0.0.0.0:0: listen tcp4 0.0.0.0:4505: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.

### ISSUE THREAD

**Issue by djfoxer** [docker/for-win#14407]
### Description

Hi, after update to 4.35.1 we discover new issue on Docker Desktop. It starts checking ports that are not blocked but are used to logging communication (listen on 4505). On 4.35.1, when we do "docker run" it throws
`Error response from daemon: Ports are not available: exposing port TCP 0.0.0.0:4505 -> 0.0.0.0:0: listen tcp4 0.0.0.0:4505: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.`
because we listen on that port for logs (log listener). Right now, after 4.35.1, we have to close our log listener then do "docker run" and after that open log listener again.

### Reproduce

docker run -p 4505:4505 (...)

### docker version

Client:
 Version:           27.3.1
 API version:       1.47
 Go version:        go1.22.7
 Git commit:        ce12230
 Built:             Fri Sep 20 11:42:27 2024
 OS/Arch:           windows/amd64
 Context:           desktop-linux

Server: Docker Desktop 4.35.1 (173168)
 Engine:
  Version:          27.3.1
  API version:      1.47 (minimum version 1.24)
  Go version:       go1.22.7
  Git commit:       41ca978
  Built:            Fri Sep 20 11:41:11 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.21
  GitCommit:        472731909fa34bd7bc9c087e4c27943f9835f111
 runc:
  Version:          1.1.13
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

### docker info

Client:
 Version:    27.3.1
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.17.1-desktop.1
    Path:     C:\Program Files\Docker\cli-plugins\docker-buildx.exe
  compose: Docker Compose (Docker Inc.)
    Version:  v2.29.7-desktop.1
    Path:     C:\Program Files\Docker\cli-plugins\docker-compose.exe
  debug: Get a shell into any image or container (Docker Inc.)
    Version:  0.0.37
    Path:     C:\Program Files\Docker\cli-plugins\docker-debug.exe
  desktop: Docker Desktop commands (Alpha) (Docker Inc.)
    Version:  v0.0.15
    Path:     C:\Program Files\Docker\cli-plugins\docker-desktop.exe
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.2
    Path:     C:\Program Files\Docker\cli-plugins\docker-dev.exe
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.27
    Path:     C:\Program Files\Docker\cli-plugins\docker-extension.exe
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.5
    Path:     C:\Program Files\Docker\cli-plugins\docker-feedback.exe
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.3.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-init.exe
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-sbom.exe
  scout: Docker Scout (Docker Inc.)
    Version:  v1.14.0
    Path:     C:\Program Files\Docker\cli-plugins\docker-scout.exe

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: 27.3.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 nvidia runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 472731909fa34bd7bc9c087e4c27943f9835f111
 runc version: v1.1.13-0-g58aa920
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
 Kernel Version: 5.15.153.1-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 20
 Total Memory: 961.6MiB
 Name: docker-desktop
 ID: b7282c14-6ff1-4e62-bdf9-e8dfb496776e
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Labels:
  com.docker.desktop.address=npipe://\\.\pipe\docker_cli
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
WARNING: daemon is not using the default seccomp profile

### Diagnostics ID

135BD9DF-E3A2-42BF-BDA1-FD684D4C0932/20241106080617

**Comment by piotrp** [comment]
For me it refused to use ports 8123 and 9000:
```sh
$ docker run -p8123:80 nginx
docker: Error response from daemon: driver failed programming external connectivity on endpoint happy_ellis (c0ee2bc965c0f5f18814e0f98fa848c828021cdafd7a4ce1a3147933d26a488c): Bind for 0.0.0.0:8123 failed: port is already allocated.

$ docker run -p9000:80 nginx
docker: Error response from daemon: driver failed programming external connectivity on endpoint gracious_roentgen (d5a01ede5ddc7cfa9c5d69c0794daef729cc93fc258c7f8287057b27149d9067): Bind for 0.0.0.0:9000 failed: port is already allocated.
```

These ports are used by a ClickHouse container, and I'm almost sure that it was running (or was created but stopped) when Docker Desktop was updated from 4.35.0.

I reinstalled Docker Desktop, seems like this helped.

Diagnostics ID: C9C5CB37-A277-44DF-9745-4AFF12CE38E6/20241106195748

**Comment by BoGnY** [comment]
should be fixed in 4.36.0 #14384

**Comment by akerouanton** [comment]
Thanks @BoGnY.

Indeed this should be fixed in v4.36 but I'm a bit confused by your original report @djfoxer:

> On 4.35.1, when we do "docker run" it throws
Error response from daemon: Ports are not available: exposing port TCP 0.0.0.0:4505 -> 0.0.0.0:0: listen tcp4 0.0.0.0:4505: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.
because we listen on that port for logs (log listener). 

From what I get, you're trying to create a port-mapping with a host port that is already in use by your log listener. In that case, how do you connect to your log listener and to the mapped port independently?

**Comment by BoGnY** [comment]
I think is something related to a false-positive already binded port.. In my case I have problem with port 53..
If I run this on a bugged docker version, I get: 
```powershell
PS C:\> Get-Process -Id (Get-NetUDPEndpoint -LocalPort 53).OwningProcess
Handles  NPM(K)    PM(K)      WS(K)     CPU(s)     Id  SI ProcessName
-------  ------    -----      -----     ------     --  -- -----------
    245      16    11136      10264       0,03   4608   0 svchost
```

But If I run this on a previous not bugged version, I get this:
```powershell
PS C:\> Get-Process -Id (Get-NetUDPEndpoint -LocalPort 53).OwningProcess
Handles  NPM(K)    PM(K)      WS(K)     CPU(s)     Id  SI ProcessName
-------  ------    -----      -----     ------     --  -- -----------
    970      48    77188      81736      21,48  13780   1 com.docker.backend
    245      16    11136      10264       0,03   4608   0 svchost
```

**Comment by akerouanton** [comment]
Port 53/udp is a bit special because that `svchost` is actually the ICS Windows service which is required by Docker Desktop for some reason.

We started working on a port-mapper enhancement that aims at improving IPv6 support. It's not totally ready yet, and thus we decided to keep it disabled for now (although it shipped in v4.35). By doing so, we introduced this regression in the 'old' port-mapper.

The issue is, the current port-mapper is using dual-stack sockets (ie. v4 and v6-capable) to bind IPv4 `0.0.0.0`. That's a divergence from Docker Engine's behavior, and this prevents separate bindings of v4 and v6 ports (eg. if you want two containers to bind the same port each but on a different address family).

So we reverted to the previous behavior, but we didn't fix the new port-mapper yet. That's mostly why I was asking about colliding port bindings in my previous comment.

Currently this works because some BSD sockets implementors (like macOS and Windows, but not Linux) allows dual-stack sockets to collide with IPv4 ports already in use. In that case, the IPv4 port will still be reachable through their IPv4 addresses (eg. `127.0.0.1`), and the dual-stack port binding will be accessible through the IPv6 loopback address (eg. `::1`), but not through the IPv4 address. When `localhost` is used, the OS will resolve that hostname to `::1` (because IPv6 addrs take precedence over IPv4).

Once the software binding the IPv4 port gets stopped, the dual-stack socket will start receiving IPv4 traffic. And trying to start again the original v4 software will fail with `address already in use`.

IMO Windows' and macOS' semantics is wrong -- it should just fail if the v4 port is already in use, and then it's developers' responsibility to handle that however they want (eg. retry in v4-only mode). But that's not what we have.

After thinking more about this issue, I'm still genuinely interested by the use cases that leverage this OS 'bug'. But I think we'll need to overload the Engine's semantics for the `HostIP` (the first field in the `-p` triplet, eg. `<host-ip>:<host-port>:<container-port>`) such that:

- When it's not specified, default to the old behavior -- bind the port in dual-stack mode. This will prevent any subsequent use of that IPv6 port.
- When it's specified to `0.0.0.0` bind in v4-only mode.
- When it's specified to `::` bind in v6-only mode.
